### ISSUE TITLE
test_photo failures at NERSC

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#2138]
@moustakas we're getting test_photo failures at NERSC, e.g. see https://data.desi.lbl.gov/desi/spectro/redux/dailytest/log/perlmutter/desispec.log .  When run as part of the full test suite, I'm getting one failure like
```
...
        if legacysurveydir is None:
            legacysurveydir = os.path.join(desi_root, 'external', 'legacysurvey', 'dr9')
    
        if not os.path.isdir(legacysurveydir):
            errmsg = f'Legacy Surveys directory {legacysurveydir} not found.'
            log.critical(errmsg)
>           raise IOError(errmsg)
E           OSError: Legacy Surveys directory /tmp/tmpgio7la0q/external/legacysurvey/dr9 not found.
```
and two failures like
```
        fiberfile = os.path.join(fiberassign_dir, stileid[:3], f'fiberassign-{stileid}.fits.gz')
        if not os.path.isfile(fiberfile):
            fiberfile = fiberfile.replace('.gz', '')
            if not os.path.isfile(fiberfile):
                errmsg = f'Fiber assignment file {fiberfile} not found!'
                log.critical(errmsg)
>               raise IOError(errmsg)
E               OSError: Fiber assignment file /tmp/tmpgio7la0q/target/fiberassign/tiles/trunk/000/fiberassign-000019.fits not found!
```

It looks like some other test might be resetting $DESI_ROOT or $DESI_ROOT_READONLY out from under you.  test_io does some $DESI_ROOT manipulation, but it also attempts to set it back when done.

But if I run `pytest py/desispec/test/test_photo.py` by itself, I get an error
```
        for tileid in truedirs.keys():
            targetdirs = gather_targetdirs(int(tileid))
            #print(tileid, targetdirs, truedirs[tileid])
>           self.assertTrue(np.all(targetdirs == truedirs[tileid]))
E           AssertionError: False is not true

py/desispec/test/test_photo.py:82: AssertionError
```

Please try running `pytest py/desispec/test` at NERSC and see if you can identify what is going wrong.  Thanks.

**Comment by sbailey** [comment]
PR #2154 fixed the individual test_photo failure, but not the problem of some other test changing DESI_ROOT without cleaning up, so I'm reopening this ticket until that is identified and fixed.

**Comment by moustakas** [comment]
I can reproduce the crash with this command (after setting the logger level to `DEBUG`), which clearly shows that `get_desi_root_readonly()` is using the temporary, cached environment set by `test/test_io.py` (bad!).
```
pytest py/desispec/test -k "TestPhoto or TestIO"
======================================================================================= test session starts =======================================================================================
platform linux -- Python 3.10.8, pytest-7.2.0, pluggy-1.0.0
rootdir: /global/u2/i/ioannis/code/desihub/desispec
plugins: astropy-header-0.1.2, filter-subpackage-0.1.1, openfiles-0.5.0, doctestplus-0.12.1, arraydiff-0.3, anyio-3.6.2, astropy-0.10.0, asdf-2.14.3, cov-4.0.0, remotedata-0.4.0, hypothesis-6.62.0, mock-3.10.0
collected 256 items / 207 deselected / 49 selected

py/desispec/test/test_io.py .............................................                                                                                                                   [ 91%]
py/desispec/test/test_photo.py F
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured stdout >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
DEBUG:meta.py:32:get_desi_root_readonly: Using cached _desi_root_readonly=/tmp/tmp9mncxdfk
DEBUG:meta.py:32:get_desi_root_readonly: Using cached _desi_root_readonly=/tmp/tmp9mncxdfk
CRITICAL:photo.py:57:gather_targetdirs: Fiber assignment file /tmp/tmp9mncxdfk/target/fiberassign/tiles/trunk/080/fiberassign-080613.fits not found!
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured log >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
[snip]
```

However, if I call just `test/test_io.py` and then check the output of `get_desi_root_readonly()`, then all is well with the world. So something about set-up vs break-down isn't happening correctly.

```
pytest py/desispec/test/test_io.py ; python -c "from desispec.io.meta import get_desi_root_readonly ; print(get_desi_root_readonly())"
======================================================================================= test session starts =======================================================================================
platform linux -- Python 3.10.8, pytest-7.2.0, pluggy-1.0.0
rootdir: /global/u2/i/ioannis/code/desihub/desispec
plugins: astropy-header-0.1.2, filter-subpackage-0.1.1, openfiles-0.5.0, doctestplus-0.12.1, arraydiff-0.3, anyio-3.6.2, astropy-0.10.0, asdf-2.14.3, cov-4.0.0, remotedata-0.4.0, hypothesis-6.62.0, mock-3.10.0
collected 45 items

py/desispec/test/test_io.py .............................................                                                                                                                   [100%]

======================================================================================= 45 passed in 3.72s ========================================================================================
DEBUG:meta.py:37:get_desi_root_readonly: Using $DESI_ROOT_READONLY=/dvs_ro/cfs/cdirs/desi
/dvs_ro/cfs/cdirs/desi
```

**Comment by moustakas** [comment]
I pushed a fix to #2155 (https://github.com/desihub/desispec/pull/2155/commits/5a9fbe580a788a3bb750a3bd9208d5e48ff4b939#diff-07c7fe343eaaae02ff9925dbc78ba75d8ce1cc483d796ce00ea97d1bd34fe9fcL89).

The issue was that `get_desi_root_readonly` (used by `test_io` but then subsequently used by `test_photo`) was caching one of the temporary directories used by `test_io` and that cache was persistent across all the `pytest`s. 

To fix this, in the `test_io.tearDownClass` class I just reset `io.meta._desi_root_readonly` back to `None`.

Once #2155 is merged (pending @weaverba137's review), we should be able to close this ticket.

```
% pytest py/desispec/test -k "TestIO or TestPhoto" -v
======================================================================================= test session starts =======================================================================================
platform linux -- Python 3.10.8, pytest-7.2.0, pluggy-1.0.0 -- /global/common/software/desi/perlmutter/desiconda/20230111-2.1.0/conda/bin/python3.10
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/global/u2/i/ioannis/code/desihub/desispec/.hypothesis/examples')
rootdir: /global/u2/i/ioannis/code/desihub/desispec
plugins: astropy-header-0.1.2, filter-subpackage-0.1.1, openfiles-0.5.0, doctestplus-0.12.1, arraydiff-0.3, anyio-3.6.2, astropy-0.10.0, asdf-2.14.3, cov-4.0.0, remotedata-0.4.0, hypothesis-6.62.0, mock-3.10.0
collected 256 items / 207 deselected / 49 selected

py/desispec/test/test_io.py::TestIO::test_addkeys PASSED                                                                                                                                    [  2%]
py/desispec/test/test_io.py::TestIO::test_backup_filename PASSED                                                                                                                            [  4%]
py/desispec/test/test_io.py::TestIO::test_camword_intersection PASSED                                                                                                                       [  6%]
py/desispec/test/test_io.py::TestIO::test_camword_spectros PASSED                                                                                                                           [  8%]
py/desispec/test/test_io.py::TestIO::test_camword_union PASSED                                                                                                                              [ 10%]
py/desispec/test/test_io.py::TestIO::test_checkzip PASSED                                                                                                                                   [ 12%]
py/desispec/test/test_io.py::TestIO::test_create_camword PASSED                                                                                                                             [ 14%]
py/desispec/test/test_io.py::TestIO::test_decode_camword PASSED                                                                                                                             [ 16%]
py/desispec/test/test_io.py::TestIO::test_desi_root_readonly PASSED                                                                                                                         [ 18%]
py/desispec/test/test_io.py::TestIO::test_dict2ndarray PASSED                                                                                                                               [ 20%]
py/desispec/test/test_io.py::TestIO::test_difference_camwords PASSED                                                                                                                        [ 22%]
py/desispec/test/test_io.py::TestIO::test_download PASSED                                                                                                                                   [ 24%]
py/desispec/test/test_io.py::TestIO::test_empty_fibermap PASSED                                                                                                                             [ 26%]
py/desispec/test/test_io.py::TestIO::test_fiberflat_rw PASSED                                                                                                                               [ 28%]
py/desispec/test/test_io.py::TestIO::test_fibermap_rw PASSED                                                                                                                                [ 30%]
py/desispec/test/test_io.py::TestIO::test_find_exposure_night PASSED                                                                                                                        [ 32%]
py/desispec/test/test_io.py::TestIO::test_find_fibermap PASSED                                                                                                                              [ 34%]
py/desispec/test/test_io.py::TestIO::test_findfile PASSED                                                                                                                                   [ 36%]
py/desispec/test/test_io.py::TestIO::test_findfile_outdir PASSED                                                                                                                            [ 38%]
py/desispec/test/test_io.py::TestIO::test_findfile_readonly PASSED                                                                                                                          [ 40%]
py/desispec/test/test_io.py::TestIO::test_fitsheader PASSED                                                                                                                                 [ 42%]
py/desispec/test/test_io.py::TestIO::test_fluxcalib PASSED                                                                                                                                  [ 44%]
py/desispec/test/test_io.py::TestIO::test_frame_rw PASSED                                                                                                                                   [ 46%]
py/desispec/test/test_io.py::TestIO::test_get_log_pathname PASSED                                                                                                                           [ 48%]
py/desispec/test/test_io.py::TestIO::test_get_nights PASSED                                                                                                                                 [ 51%]
py/desispec/test/test_io.py::TestIO::test_get_reduced_frames PASSED                                                                                                                         [ 53%]
py/desispec/test/test_io.py::TestIO::test_get_tempfilename PASSED                                                                                                                           [ 55%]
py/desispec/test/test_io.py::TestIO::test_image_rw PASSED                                                                                                                                   [ 57%]
py/desispec/test/test_io.py::TestIO::test_io_qa_frame PASSED                                                                                                                                [ 59%]
py/desispec/test/test_io.py::TestIO::test_iotime PASSED                                                                                                                                     [ 61%]
py/desispec/test/test_io.py::TestIO::test_native_endian PASSED                                                                                                                              [ 63%]
py/desispec/test/test_io.py::TestIO::test_parse_cameras PASSED                                                                                                                              [ 65%]
py/desispec/test/test_io.py::TestIO::test_read_table PASSED                                                                                                                                 [ 67%]
py/desispec/test/test_io.py::TestIO::test_readonly_filepath PASSED                                                                                                                          [ 69%]
py/desispec/test/test_io.py::TestIO::test_relsymlink PASSED                                                                                                                                 [ 71%]
py/desispec/test/test_io.py::TestIO::test_replace_prefix PASSED                                                                                                                             [ 73%]
py/desispec/test/test_io.py::TestIO::test_search_framefile PASSED                                                                                                                           [ 75%]
py/desispec/test/test_io.py::TestIO::test_shorten_filename PASSED                                                                                                                           [ 77%]
py/desispec/test/test_io.py::TestIO::test_sky_rw PASSED                                                                                                                                     [ 79%]
py/desispec/test/test_io.py::TestIO::test_skycorr_pca_rw PASSED                                                                                                                             [ 81%]
py/desispec/test/test_io.py::TestIO::test_skycorr_rw PASSED                                                                                                                                 [ 83%]
py/desispec/test/test_io.py::TestIO::test_stdstar PASSED                                                                                                                                    [ 85%]
py/desispec/test/test_io.py::TestIO::test_supports_memmap PASSED                                                                                                                            [ 87%]
py/desispec/test/test_io.py::TestIO::test_sv1_faflavor2program PASSED                                                                                                                       [ 89%]
py/desispec/test/test_io.py::TestIO::test_write_bintable PASSED                                                                                                                             [ 91%]
py/desispec/test/test_photo.py::TestPhoto::test_gather_targetdirs PASSED                                                                                                                    [ 93%]
py/desispec/test/test_photo.py::TestPhoto::test_gather_targetphot PASSED                                                                                                                    [ 95%]
py/desispec/test/test_photo.py::TestPhoto::test_gather_tractorphot PASSED                                                                                                                   [ 97%]
py/desispec/test/test_photo.py::TestPhoto::test_gather_tractorphot_dr10 PASSED                                                                                                              [100%]

=============================================================================== 49 passed, 207 deselected in 24.48s ===============================================================================
```

**Comment by sbailey** [comment]
Good catch.  Thanks for the fix.  I confirm that after merging #2155, the full desispec test suite passes at NERSC.  Closing.
